src/cudadecoder/batched-threaded-nnet3-cuda-online-pipeline.cc

```diff
@@ -0,0 +1,375 @@
+  list_channels_last_chunk_.clear();
+  list_corr_id_last_chunk_.clear();
+  for (int i = 0; i < is_last_chunk.size(); ++i) {
+    if (is_last_chunk[i]) {
+      list_channels_last_chunk_.push_back(channels[i]);
+      list_corr_id_last_chunk_.push_back(corr_ids[i]);
+    }
+  }
+  cuda_decoder_->PrepareForGetRawLattice(list_channels_last_chunk_, true);
+  // Storing number of callbacks not done. Used if  WaitForLatticeCallbacks()
+  // is
+  // called
+  n_lattice_callbacks_not_done_.fetch_add(list_channels_last_chunk_.size(),
+                                          std::memory_order_acquire);
+
+  // delete data used for decoding that corr_id
+  for (int32 i = 0; i < list_channels_last_chunk_.size(); ++i) {
+    uint64_t ichannel = list_channels_last_chunk_[i];
+    CorrelationID corr_id = list_corr_id_last_chunk_[i];
+    int32 ndeleted = corr_id2channel_.erase(corr_id);
+    KALDI_ASSERT(ndeleted == 1);
+    KALDI_ASSERT(thread_pool_->tryPush(
+        {&BatchedThreadedNnet3CudaOnlinePipeline::FinalizeDecodingWrapper, this,
+         ichannel, corr_id}));
+  }
+  list_channels_last_chunk_.clear();
+  list_corr_id_last_chunk_.clear();
+}
+
+void BatchedThreadedNnet3CudaOnlinePipeline::ListIChannelsInBatch(
+    const std::vector<CorrelationID> &corr_ids, std::vector<int> *channels) {
+  channels->clear();
+  list_channels_last_chunk_.clear();
+  list_corr_id_last_chunk_.clear();
+  for (int i = 0; i < corr_ids.size(); ++i) {
+    int corr_id = corr_ids[i];
+    auto it = corr_id2channel_.find(corr_id);
+    KALDI_ASSERT(it != corr_id2channel_.end());
+    int ichannel = it->second;
+    channels->push_back(ichannel);
+  }
+}
+
+void BatchedThreadedNnet3CudaOnlinePipeline::RunNnet3(
+    const std::vector<int> &channels,
+    const std::vector<BaseFloat *> &d_features, const int features_stride,
+    const std::vector<int> &n_input_frames_valid,
+    const std::vector<bool> &is_last_chunk,
+    const std::vector<BaseFloat *> &d_ivectors) {
+  cuda_nnet3_->RunBatch(channels, d_features, features_stride, d_ivectors,
+                        n_input_frames_valid, is_last_chunk,
+                        &d_all_log_posteriors_, &all_frames_log_posteriors_);
+}
+
+void BatchedThreadedNnet3CudaOnlinePipeline::RunDecoder(
+    const std::vector<int> &channels) {
+  for (int iframe = 0; iframe < all_frames_log_posteriors_.size(); ++iframe) {
+    cuda_decoder_->AdvanceDecoding(all_frames_log_posteriors_[iframe]);
+  }
+}
+
+void BatchedThreadedNnet3CudaOnlinePipeline::ReadParametersFromModel() {
+  // TODO remove that
+  feature_info_.reset(new OnlineNnet2FeaturePipelineInfo(config_.feature_opts));
+  feature_info_->ivector_extractor_info.use_most_recent_ivector = true;
+  feature_info_->ivector_extractor_info.greedy_ivector_extractor = true;
+
+  OnlineNnet2FeaturePipeline feature(*feature_info_);
+  // TODO clean following lines
+  input_dim_ = feature.InputFeature()->Dim();
+  ivector_dim_ = feature.IvectorFeature()->Dim();
```

Comment:
Does the current code require a MFCC+ivector model? I tried running `batched-wav-nnet3-cuda` with a model on hand that uses fbank without ivectors, and it SegFaults at this line.

Reply:
my fbank-based model works fine (using batched-wav-nnet3-cuda executable).

Reply:
Yes, it's not tested yet without ivectors